File: AtCoder/ABC/172/python/C_Tsundoku.py
from bisect import bisect

INF = float('inf')


N, M, K = map(int, input().split())
A = list(map(int, input().split()))
B = list(map(int, input().split()))

sum_A = [0] * (N + 1)
for i, a in enumerate(A):
    sum_A[i + 1] = sum_A[i] + a
sum_B = [0] * (M + 1) + [INF]
for i, b in enumerate(B):
    sum_B[i + 1] = sum_B[i] + b

res = 0
for i in range(N + 1):
    if K - sum_A[i] >= 0:
        j = bisect(sum_B, K - sum_A[i])
        res = max(res, i + j - 1)
print(res)

File: AtCoder/ABC/172/cpp/C_Tsundoku.cpp
#include <bits/stdc++.h>
#define MAX_N 200'000
#define MAX_M 200'000
#define MAX_K 1'000'000'000
using namespace std;
typedef long long ll;

int N, M, K;
ll A[MAX_N], B[MAX_M];
ll sum_A[MAX_N + 1], sum_B[MAX_M + 2];

void solve() {
  for (int i = 0; i < N; ++i) {
    sum_A[i + 1] = sum_A[i] + A[i];
  }
  for (int i = 0; i < M; ++i) {
    sum_B[i + 1] = sum_B[i] + B[i];
  }
  sum_B[M + 1] = MAX_K + 1;
  int res = 0;
  for (int i = 0; i <= N; ++i) {
    if (K - sum_A[i] >= 0) {
      int j = upper_bound(sum_B, sum_B + M + 1, K - sum_A[i]) - sum_B;
      res = max(res, i + j - 1);
    }
  }
  printf("%d\n", res);
}

int main() {
  cin >> N >> M >> K;
  for (int i = 0; i < N; ++i) {
    cin >> A[i];
  }
  for (int i = 0; i < M; ++i) {
    cin >> B[i];
  }
  solve();
}